lib/test_queue/runner.rb

```diff
@@ -427,40 +427,53 @@ def distribute_queue
         else
           sock = @server.accept
           cmd = sock.gets.strip
+
+          cmd =~ /^TOKEN=(\w+)/
+          # If we have a slave from a different test run, don't respond, and it will consider the test run done.
+          case $1
+          # No special handling if the token is valid
+          when @run_token
+          when nil
+            STDERR.puts "*** Worker sent no token to master for run #{@run_token}; ignoring."
+            sock.write("WRONG RUN\n")
+            File.open("/tmp/errors.log", "a") {|f| f.puts "WRONG RUN - nil token"}
+            next
+          else
+            STDERR.puts "*** Worker from run #{run_token} connected to master for run #{@run_token}; ignoring."
+            sock.write("WRONG RUN\n")
+            File.open("/tmp/errors.log", "a") {|f| f.puts "WRONG RUN - wrong token"}
+            next
+          end
+
           case cmd
-          when /^POP/
+          when /^TOKEN=\w+ POP/
```

Comment:
Maybe we should mutate `cmd` here to strip the token to remove the repetition afterwards? Or at least call the original `message` or something, strip the token, and call the rest `cmd`.